docs/rfcs/025-generation-numbers.md

```diff
@@ -292,125 +299,400 @@ the index storage format is migrated to a binary format from JSON.
 
 ### `pageserver::Tenant` startup changes
 
-#### Attachment
-
-We already reconcile with remote metadata during attachment. It is necessary to synchronize
-with remote metadata even if there is some local metadata that could be brought up to
-date by ingesting the WAL, because if another node updated
-the remote storage, it might have advanced the remote_consistent_lsn such that the safekeeper drops WAL
-content that would be needed for this node to recover. Reconcilation with remote metadata also acts as an optimization, to avoid redundantly
-re-ingesting WAL data that is already in S3.
+#### Metadata reconciliation
+
+Whether for a new attachment or just a pageserver restart, Tenant _should_
+reconcile with remote metadata.
+
+If the remote index reaches a lower LSN than the pageserver's local state,
+then the pageserver is free to ignore the remote index and use its local
+state as-is.
+
+If the remote index reaches a higher LSN than the pageserver's local state,
+then this indicates that the local state may not be safely used, unless the
+safekeepers still have the WAL that spans the range between the local state
+and the remote state. In this case, the pageserver may choose between replaying
+the WAL to bring its local state up to date, or reconciling its local state with
+the contents of S3 without replaying the WAL again:
```

Comment:
> In this case, the pageserver may choose between replaying
> the WAL to bring its local state up to date, or reconciling its local state with
> the contents of S3 without replaying the WAL again:

Before this RFC, the above was highly unsafe. Would be worth emphasizing this early in above paragraph.